File: CSharp/Supermarket/ProductRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Supermarket.Model.Core;

namespace Supermarket
{
    public class ProductRepository
    {
        public ProductRepository()
        {
            Products = GetProductList();
        }

        public List<Product> Products { get; }

        private List<Product> GetProductList()
        {
            const int productCount = 15;

            return Enumerable.Range(1, productCount)
                             .Select(CreateAProduct)
                             .ToList();
        }

        private static Product CreateAProduct(int i)
        {
            var dicount = i%3;
            var vat = i%2;
            var product = new Product
            {
                Name = $"Product {i}",
                UnitPrice = i*11,
                Discount = dicount == 0 ? Discount.None : dicount == 1 ? Discount.Take3Pay2 : Discount.Take5Pay4,
                VATPercentage = vat == 0 ? 10 : 20
            };
            return product;
        }

        public Product GetProductByNumber(EAN number)
        {
            if (number == null)
            {
                throw new ArgumentNullException(nameof(number));
            }
            return Products.SingleOrDefault(item => item.Number.Code == number.Code);
        }

        public void Add(Product product)
        {
            Products.Add(product);
        }
    }
}

File: java/src/main/java/supermarket/ProductRepository.java
package supermarket;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import supermarket.model.core.EAN;
import supermarket.model.core.Product;

public class ProductRepository implements IProductRepository {
	private List<Product> products;

	public ProductRepository() {
		this.products = new ArrayList<Product>();
	}

	public Product getProductByNumber(final EAN number) {
		if (number == null) {
			throw new IllegalArgumentException();
		}
		return products.stream()//
				.filter(product -> number.equals(product.getNumber()))//
				.findAny()//
				.orElse(null);
	}

	public Product getProductByName(String name) {
		return products.stream() //
				.filter(product -> name.equals(product.getName()))//
				.findAny()//
				.orElse(null);
	}

	public void add(Product product) {
		products.add(product);
	}

	public void addAll(Collection<Product> additionalProducts) {
		products.addAll(additionalProducts);
	}

}
